crates/wasi-c2/src/file.rs

```diff
@@ -0,0 +1,246 @@
+    pub const ALLOCATE: Self = FileCaps { flags: 256 };
+    pub const FILESTAT_GET: Self = FileCaps { flags: 512 };
+    pub const FILESTAT_SET_SIZE: Self = FileCaps { flags: 1024 };
+    pub const FILESTAT_SET_TIMES: Self = FileCaps { flags: 2048 };
+}
+
+impl std::ops::BitOr for FileCaps {
+    type Output = FileCaps;
+    fn bitor(self, rhs: FileCaps) -> FileCaps {
+        FileCaps {
+            flags: self.flags | rhs.flags,
+        }
+    }
+}
+
+impl std::fmt::Display for FileCaps {
+    fn fmt(&self, _f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        todo!()
+    }
+}
+
+pub struct FdStat {
+    pub filetype: Filetype,
+    pub base_caps: FileCaps,
+    pub inheriting_caps: FileCaps,
+    pub flags: FdFlags,
+}
+
+impl WasiFile for cap_std::fs::File {
+    fn datasync(&self) -> Result<(), Error> {
+        self.sync_data()?;
+        Ok(())
+    }
+    fn sync(&self) -> Result<(), Error> {
+        self.sync_all()?;
+        Ok(())
+    }
+    fn get_filetype(&self) -> Result<Filetype, Error> {
+        let meta = self.metadata()?;
+        // cap-std's Metadata/FileType only offers booleans indicating whether a file is a directory,
+        // symlink, or regular file.
+        // Directories should be excluded by the type system.
+        if meta.is_file() {
+            Ok(Filetype::RegularFile)
+        } else {
+            Err(Error::Badf) // XXX idk what to do here
+        }
+    }
+    fn get_fdflags(&self) -> Result<FdFlags, Error> {
+        // XXX cap-std doesnt expose append, dsync, nonblock, rsync, sync
+        todo!()
+    }
+    fn get_oflags(&self) -> Result<OFlags, Error> {
+        // XXX what if it was opened append, async, nonblock...
+        let perms = self.metadata()?.permissions();
+        if perms.readonly() {
+            Ok(OFlags::RDONLY)
+        } else {
+            Ok(OFlags::RDWR)
+        }
+    }
+    fn set_oflags(&self, _flags: OFlags) -> Result<(), Error> {
+        // XXX cap-std::fs::Permissions does not export a constructor to build this out of
+        #[allow(unreachable_code)]
+        self.set_permissions(todo!())?;
+        Ok(())
+    }
+    fn get_filestat(&self) -> Result<Filestat, Error> {
+        // XXX cap-std does not expose every part of filestat
+        #![allow(unreachable_code, unused_variables)]
+        let meta = self.metadata()?;
+        Ok(Filestat {
+            device_id: todo!(),
+            inode: todo!(),
+            filetype: self.get_filetype()?,
+            nlink: todo!(),
+            size: meta.len(),
+            atim: meta.accessed()?.into_std(),
+            mtim: meta.modified()?.into_std(),
+            ctim: meta.created()?.into_std(),
+        })
+    }
+    fn set_filestat_times(
+        &self,
+        _atim: Option<FilestatSetTime>,
+        _mtim: Option<FilestatSetTime>,
+    ) -> Result<(), Error> {
+        // XXX cap-std does not expose a way to set accessed time or modified time
```

Comment:
Ah, sorry I neglected to mention it, that's in a separate crate -- [fs-set-times](https://crates.io/crates/fs-set-times). If you use that crate's [SystemTimeSpec](https://docs.rs/fs-set-times/0.2.1/fs_set_times/enum.SystemTimeSpec.html) in place the `FilestatSetTime` definition above, then usage is basically

```rust
use fs_set_times::SetTimes;
file.set_times(atime, mtime)
```

Or something like `pub trait WasiFile: FileIoExt + SetTimes` might work too.